timemachine/potentials/jax_utils.py

```diff
@@ -44,29 +44,15 @@ def delta_r(ri, rj, box=None):
 
     return diff
 
-# def distance(ri, rj, box=None, gij=None):
-#     # assert box is None
-#     if gij is not None:
-#         deltas_4d = np.power(ri - rj, 2)   
-#         # print(deltas_4d.shape)
-#         deltas_3d = deltas_4d[..., :3]
-#         # print(deltas_3d.shape)
-#         dij_4d = np.sqrt(np.sum(deltas_4d, axis=-1))
-#         dij_3d = np.sqrt(np.sum(deltas_3d, axis=-1))
 
-#         # print("shapes", gij.shape, dij_3d.shape, dij_4d.shape)
-#         dij = np.where(gij, dij_3d, dij_4d)
-#     else:
-#         deltas = np.power(ri - rj, 2)
-#         dij = np.sqrt(np.sum(deltas, axis=-1))
-
-#     # print(dij)
-
-#     return dij
-
-def distance(ri, rj, box=None):
-    dxdydz = np.power(delta_r(ri, rj, box), 2)
-    # np.linalg.norm nans but this doesn't
-    dij = np.sqrt(np.sum(dxdydz, axis=-1))
+def distance(x, box):
+    # nonbonded distances require the periodic box
+    assert x.shape[1] == 3 or x.shape[1] == 4 # 3d or 4d
+    ri = np.expand_dims(x, 0)
+    rj = np.expand_dims(x, 1)
+    d2ij = np.sum(np.power(delta_r(ri, rj, box), 2), axis=-1)
+    N = d2ij.shape[0]
+    d2ij = np.where(np.eye(N), 0, d2ij)
+    dij = np.where(np.eye(N), 0, np.sqrt(d2ij))
```

Comment:
Question: would it make sense to mirror the distinction between `pdist` and `cdist` in scipy.spatial.distance here? https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cdist.html

(Sometimes we're interested in distances just between one subset of atoms and another, for example when doing something different with ligand-protein distances than with the rest of the distances.)

Reply:
I think so, but I'd prefer to properly implement ```pdist/cdist``` in a separate pass to avoid introducing too many changes at once.

Reply:
(This PR is not intended to extend functionality)